---
title: "Differential Expression"
author: "Harvard Chan Bioinformatics Core"
date: "`r Sys.Date()`"
output:
   html_document:
      code_folding: hide
      df_print: paged
      highlights: pygments
      number_sections: true
      self_contained: true
      theme: default
      toc: true
      toc_float:
         collapsed: true
         smooth_scroll: true
editor_options: 
  chunk_output_type: console
params:
  numerator: foo
  denominator: bar
  subset_value: nah
  params_file: params_de.R

---

```{r echo = F}
source(params$params_file)
```

```{r, cache = FALSE, message = FALSE, warning=FALSE, echo=FALSE,}
library(rtracklayer)
library(DESeq2)
library(tidyverse)
library(stringr)
library(RUVSeq)
library(DEGreport)
library(ggpubr)
library(msigdbr)
library(fgsea)
library(org.Hs.eg.db)
library(knitr)

ggplot2::theme_set(theme_light(base_size = 14))
opts_chunk[["set"]](
    cache = F,
    cache.lazy = FALSE,
    dev = c("png", "pdf"),
    error = TRUE,
    highlight = TRUE,
    message = FALSE,
    prompt = FALSE,
    tidy = FALSE,
    warning = FALSE,
    echo = F, 
    fig.height = 4)
```

```{r sanitize-datatable}
sanitize_datatable = function(df, ...) {
 # remove dashes which cause wrapping
 DT::datatable(df, ..., rownames=gsub("-", "_", rownames(df)),
                   colnames=gsub("-", "_", colnames(df)))
}
```

# Overview

-   Project: `r project`
-   PI: `r PI`
-   Analyst: `r analyst`
-   Experiment: `r experiment`
-   Aim: `r aim`
-   Comparison: `r paste0(params$subset_value, ': ', params$numerator, ' vs. ', params$denominator)`

```{r create-filenames}
filenames = str_interp("${params$subset_value}_${params$numerator}_vs_${params$denominator}")
contrasts = c(column,params$numerator,params$denominator)

name_expression_fn=file.path(root,
                             basedir,
                             str_interp("${filenames}_expression.csv"))
name_deg_fn=file.path(root,
                      basedir,
                      str_interp("${filenames}_deg.csv"))
name_pathways_fn=file.path(root,
                           basedir,
                           str_interp("${filenames}_pathways.csv"))

```

```{r read-counts-data}
coldata=read.csv(metadata_fn, row.names = 1)
stopifnot(column %in% names(coldata))

# use only some samples, by default use all
coldata <- coldata[coldata[[paste(subset_column)]] == params$subset_value, ]
coldata <- coldata[coldata[[paste(column)]] %in% c(params$numerator, params$denominator), ]
rownames(coldata) <- coldata$description
coldata[[column]] = relevel(as.factor(coldata[[column]]), params$denominator)

counts <- read_csv(counts_fn) %>% column_to_rownames('gene')
counts <- counts[,colnames(counts) %in% coldata$description]


# if the names don't match in order or string check files names and coldata information
counts = counts[rownames(coldata)]
stopifnot(all(names(counts) == rownames(coldata)))

# gtf_df <- as.data.frame(rtracklayer::import(GTF))
# gtf_df$gene_id <- sub("\\..*","\\", gtf_df$gene_id) # remove the ensembl numbering
# 
# rdata <- gtf_df %>% filter(type == 'gene', gene_id %in% rownames(counts)) %>%
#   dplyr::select(gene_id, gene_name)
# rdata$gene_name[is.na(rdata$gene_name)] <- rdata$gene_id[is.na(rdata$gene_name)]

rdata = AnnotationDbi::select(org.Hs.eg.db, rownames(counts), 'SYMBOL', 'ENSEMBL') %>%
  dplyr::select(gene_id = ENSEMBL, gene_name = SYMBOL)

```

# Remove Unwanted Variability

When performing differential expression analysis, it is important to ensure that any detected differences are truly a result of the experimental comparison being made and not any additional variability in the data. 

```{r before-RUV}

dds_before <- DESeqDataSetFromMatrix(counts, coldata, design = ~1)

vsd_before <- vst(dds_before)

degPCA(assay(vsd_before), colData(vsd_before), 
       condition = column) + ggtitle('Before RUV')

```

```{r do-RUV}
design <-  coldata[[column]]
names(design) <- coldata$name
diffs <- makeGroups(design)
dat <- assay(vsd_before)
ruvset <- RUVs(dat, cIdx=rownames(dat), k=1, diffs, isLog = T, round = F)
vars <- ruvset$W

new_cdata <- cbind(coldata, vars)

formula <- as.formula(paste0("~ ", 
                          paste0(
                            colnames(new_cdata)[grepl("W", colnames(new_cdata))], 
                            collapse = " + "
                          ), " + ", column)
)
```

```{r after-RUV}
## Check if sample name matches
stopifnot(all(names(counts) == rownames(new_cdata)))

dds_after <- DESeqDataSetFromMatrix(counts, new_cdata, design = formula)
vsd_after<- vst(dds_after, blind=FALSE)
degPCA(ruvset$normalizedCounts, new_cdata, 
       condition = column) + ggtitle('After RUV')

```


# Differential Expression

Because it is difficult to accurately detect and quantify the expression of lowly expressed genes, differences in their expression between treatment conditions can be unduly exaggerated. We correct for this so that gene LFC is not dependent overall on basal gene expression level.

```{r DE}
de <- DESeq(dds_after)

# resultsNames(de) # check the order is right
resLFC = results(de, contrast=contrasts)
resLFCS <- lfcShrink(de, coef=resultsNames(de)[ncol(vars)+2], type="apeglm")

res <- as.data.frame(resLFCS) %>%
  rownames_to_column('gene_id') %>% left_join(rdata, by = 'gene_id') %>% 
  relocate(gene_name) %>% rename(lfc = log2FoldChange) %>%
  mutate(pi = abs(lfc) * -log10(padj)) %>% arrange(-pi)

res_sig <- res %>% filter(padj < 0.05) %>% arrange(padj)

res_mod <- res %>% mutate(lfc = replace(lfc, lfc < -5, -5)) %>% mutate(lfc = replace(lfc, lfc > 5, 5))
show <- as.data.frame(res_mod[1:10, c("lfc", "padj", "gene_name")])

degMA(as.DEGSet(resLFC)) + ggtitle('Before LFC Shrinking')
```

```{r}
degMA(as.DEGSet(resLFCS), limit = 2) + ggtitle('After LFC Shrinking')

```

This volcano plot shows the genes that are significantly up- and down-regulated as a result of the difference in treatment. 
```{r}
degVolcano(res_mod[,c('lfc', 'padj')], plot_text = show) 

```

## Differentially Expressed Genes
```{r sig_genes_table}
res_sig %>% sanitize_datatable
```

# Pathway Enrichment

From the set of differentially expressed genes and using publicly available information about gene sets involved in biological processes and functions, we can calculate which biological processes and functions are significantly perturbed as a result of the treatment. 

```{r}
universe=res %>% 
  filter(!is.na(padj)) %>% pull(gene_id)
mapping = AnnotationDbi::select(org.Hs.eg.db, universe, 'ENTREZID', 'ENSEMBL')

ranks_df <- res %>% 
  filter(padj < 0.01, !is.na(padj)) %>% # only if enough DE genes  
  arrange((lfc)) %>% 
  left_join(mapping, by = c("gene_id"="ENSEMBL")) %>% 
  rename(entrez_id=ENTREZID) %>% 
  filter(!is.na(entrez_id) & !is.na(padj))
ranks <- ranks_df$lfc
names(ranks) <- ranks_df$entrez_id

all_in_life=list(
  msigdbr(species = "human", category = "H") %>% mutate(gs_subcat="Hallmark"),
  msigdbr(species = "human", category = "C2", subcategory = "CP:REACTOME"),
  msigdbr(species = "human", category = "C2", subcategory = "CP:KEGG"),
  msigdbr(species = "human", category = "C2", subcategory = "CP:PID"),
  msigdbr(species = "human", category = "C5", subcategory = "GO:BP"),
  msigdbr(species = "human", category = "C5", subcategory = "GO:MF"),
  msigdbr(species = "human", category = "C5", subcategory = "HPO"),
  msigdbr(species = "human", category = "C3", subcategory = "TFT:GTRD"),
  msigdbr(species = "human", category = "C6") %>% mutate(gs_subcat="Oncogenic")
)

ora_input = res %>% filter(!is.na(padj), padj<0.01, abs(lfc)>0.3) %>% pull(gene_id)
input_entrezid <- AnnotationDbi::select(org.Hs.eg.db, ora_input, 'ENSEMBL', columns = c('ENTREZID', 'SYMBOL'))

total_deg=length(unique(ora_input))/length(unique(mapping$ENTREZID))
pathways_ora_all = lapply(all_in_life, function(p){
  pathway = split(x = p$entrez_gene, f = p$gs_name)
  db_name = paste(p$gs_cat[1], p$gs_subcat[1],sep=":")
  respath <- fora(pathways = pathway, 
                  genes = unique(input_entrezid$ENTREZID),
                  universe = unique(mapping$ENTREZID),
                  minSize  = 15,
                  maxSize  = 500)
  coll_respath = collapsePathwaysORA(respath[order(pval)][padj < 0.1], 
                                     pathway, unique(input_entrezid$ENTREZID), unique(mapping$ENTREZID))
  as_tibble(respath[pathway %in% coll_respath$mainPathways])  %>% 
    mutate(database=db_name, NES=(overlap/size)/(total_deg))
}) %>% bind_rows() %>% 
  mutate(analysis="ORA")
  
ora_tb = pathways_ora_all %>% unnest(overlapGenes) %>%
  group_by(pathway) %>% 
  left_join(mapping, by =c("overlapGenes"="ENTREZID")) %>% 
  dplyr::select(pathway, padj, NES, ENSEMBL, analysis,
                database)

pathways_long = ora_tb

```


```{r pathaways_table}
pathways_ora_all %>% sanitize_datatable()
```


```{r write-files}
counts_norm=ruvset$normalizedCounts %>% as.data.frame() %>% 
  rownames_to_column("gene_id")
write_csv(counts_norm %>% 
            mutate(subset = params$subset_value, 
                   comparison = str_interp("${params$numerator}_vs_${params$denominator}")), 
          name_expression_fn)
write_csv(res %>% 
            mutate(subset = params$subset_value, 
                   comparison = str_interp("${params$numerator}_vs_${params$denominator}")), 
          name_deg_fn)
write_csv(pathways_long %>% 
            mutate(subset = params$subset_value, 
                   comparison = str_interp("${params$numerator}_vs_${params$denominator}")), 
          name_pathways_fn)
```
# R session

List and version of tools used for the DE report generation.

```{r}
sessionInfo()
```
