---
title: "ChIPSeq DiffBind"
author: "Harvard Chan Bioinformatics Core"
date: "`r Sys.Date()`"
output:
   html_document:
      code_folding: hide
      df_print: paged
      highlights: pygments
      number_sections: true
      self_contained: true
      theme: default
      toc: true
      toc_float:
         collapsed: true
         smooth_scroll: true
editor_options: 
  chunk_output_type: console
params:
  # Fill this file with the right paths to nfcore output
  # .qs file name for saving DiffBind Counts object
  params_file: params_diffbind-example.R
  project_file: ../information.R
  functions_file: ../libs/load_data.R
  diffbind_counts_file: diffbind_counts.qs 
  factor_of_interest: genotype
  numerator: cKO
  denominator: WT
  # species = mouse or human
  species: mouse
---

```{r, cache = FALSE, message = FALSE, warning=FALSE}
# This set up the working directory to this file so all files can be found
library(rstudioapi)
setwd(fs::path_dir(getSourceEditorContext()$path))
```


```{r source_params, cache = FALSE, message = FALSE, warning=FALSE}
# 1. set up factor_of_interest parameter from parameter above or manually
#    this is used to color plots, it needs to be part of the metadata
# 2. Set input files in this file
source(params$params_file)
# 3. If you set up this file, project information will be printed below and
#.   it can be reused for other Rmd files.
source(params$project_file)
# 4. Load custom functions to load data from coldata/metrics/counts
source(params$functions_file)
```

# Overview

-   Project: `r project`
-   PI: `r PI`
-   Analyst: `r analyst`
-   Experiment: `r experiment`


```{r load_libraries, cache = FALSE, message = FALSE, warning=FALSE}
library(tidyverse)
library(knitr)
# library(rtracklayer)
library(DESeq2)
library(DEGreport)
library(ggrepel)
# library(RColorBrewer)
library(DT)
library(pheatmap)
library(bcbioR)
library(janitor)
# library(ChIPpeakAnno)
library(UpSetR)
library(DiffBind)
library(qs)
library(EnhancedVolcano)
library(ggprism)
library(ChIPseeker)

if (params$species == 'mouse'){
  library(TxDb.Mmusculus.UCSC.mm10.knownGene)
  txdb <- TxDb.Mmusculus.UCSC.mm10.knownGene
  anno_db <- 'org.Mm.eg.db'
} else if (params$species == human){
  library(TxDb.Hsapiens.UCSC.hg38.knownGene)
  txdb <- TxDb.Hsapiens.UCSC.hg38.knownGene
  anno_db <- 'org.Hs.eg.db'
}


colors=cb_friendly_cols(1:15)
ggplot2::theme_set(theme_prism(base_size = 14))
opts_chunk[["set"]](
    cache = FALSE,
    cache.lazy = FALSE,
    dev = c("png", "pdf"),
    error = TRUE,
    highlight = TRUE,
    message = FALSE,
    prompt = FALSE,
    tidy = FALSE,
    warning = FALSE,
    fig.height = 4)

# set seed for reproducibility
set.seed(1234567890L)

```


```{r sanitize-datatable}
sanitize_datatable = function(df, ...) {
 # remove dashes which cause wrapping
 DT::datatable(df, ..., rownames=gsub("-", "_", rownames(df)),
                   colnames=gsub("-", "_", colnames(df)))
}
```

# Samples and metadata

```{r load_data, message=F, warning=F}
coldata <- load_coldata(coldata_fn)

# DiffBind requires a very specific samplesheet in order to create the peak counts object, see https://www.rdocumentation.org/packages/DiffBind/versions/2.0.2/topics/dba for further details

# make_diffbind_samplesheet is a function provided by bcbioR to help assemble DiffBind's samplesheet 
# using the nf-core samplesheet and output. In the resulting DiffBind counts object, it
# encodes your factor of interest as "Factor" and the antibody as "Condition"

samplesheet <- make_diffbind_samplesheet(coldata, bam_dir, peaks_dir, params$factor_of_interest)

# if necessary, one additional covariate of interest can be encoded as "Tissue" 

```

```{r show_metadata}
samplesheet %>% dplyr::select(SampleID, Replicate, Condition, Factor, ControlID) %>% sanitize_datatable()
```

```{r create diffbind counts object, eval = !file.exists(params$diffbind_counts_file)}

diffbind_obj <- dba(sampleSheet = samplesheet, scoreCol = 5)

# This command may take several minutes. Recommend using multiple cores and lots of memory
diffbind_counts <- dba.count(diffbind_obj, bUseSummarizeOverlaps = TRUE, bParallel = T)

# save object when time-intensive command is finished, so that this cell only need run once
qsave(diffbind_counts, params$diffbind_counts_file)

```

# PCA 
```{r PCA}

diffbind_counts <- qread(params$diffbind_counts_file)

diffbind_norm <- dba.normalize(diffbind_counts)

norm_counts <- dba.peakset(diffbind_norm, bRetrieve=TRUE, DataType=DBA_DATA_FRAME) %>%
  mutate(peak = paste(CHR, START, END, sep = '_')) %>%
  dplyr::select(-CHR, -START, -END) 
rownames(norm_counts) <- norm_counts$peak
norm_counts <- norm_counts %>% dplyr::select(-peak) %>% as.matrix()
norm_counts_log <- log2(norm_counts + 1)

coldata_for_pca <- coldata[colnames(norm_counts), ]

stopifnot(all(colnames(norm_counts) == rownames(coldata_for_pca)))

degPCA(norm_counts_log, coldata_for_pca, condition = params$factor_of_interest) +
  scale_color_cb_friendly()

```

# Differentially Bound Peaks

## Table
```{r DE analysis}
diffbind_norm <- dba.contrast(diffbind_norm, contrast = c('Factor', params$numerator, params$denominator))
results_obj <- dba.analyze(diffbind_norm, bGreylist = F)

results_report <- dba.report(results_obj, th = 1) 
results_report_sig <- dba.report(results_obj)

results <- results_report %>% as.data.frame()
results_sig <- results_report_sig %>% as.data.frame()

results_sig %>% sanitize_datatable()

```

## Volcano plot
```{r volcano, fig.height = 8}
results_mod <- results %>% 
  mutate(Fold = replace(Fold, Fold < -5, -5)) %>% 
  mutate(Fold = replace(Fold, Fold > 5, 5)) %>%
  mutate(peak = paste(seqnames, start, end, sep = '_'))
show <- as.data.frame(results_mod[1:6, c("Fold", "FDR", "peak")])
EnhancedVolcano(results_mod,
                lab= results_mod$peak, 
                pCutoff = 0.05, 
                selectLab = c(show$peak),
                FCcutoff = 0.5,
                x = 'Fold',
                y = 'FDR', 
                title = paste(params$factor_of_interest, ':', params$numerator, 'vs', params$denominator),
                col=as.vector(colors[c("dark_grey", "light_blue",
                                       "purple", "purple")]),
                subtitle = "", drawConnectors = T,  max.overlaps = Inf) 

```

## Plot top peaks
```{r plot top peaks, fig.width = 8, fig.height = 6}
norm_counts_log_long <- norm_counts_log %>% as.data.frame() %>%
  rownames_to_column('peak') %>%
  pivot_longer(!peak, names_to = 'sample', values_to = 'norm_counts_log2') %>%
  left_join(coldata)

norm_counts_log_long_top <- norm_counts_log_long %>% filter(peak %in% show$peak)

ggplot(norm_counts_log_long_top, aes(x = .data[[params$factor_of_interest]], y = norm_counts_log2)) + 
  facet_wrap(~peak, scale = 'free_y') + geom_boxplot()
```

## Annotate DB peaks

```{r annotate, echo = F}

results_sig_anno <- annotatePeak(results_report_sig, 
                                 tssRegion = c(-2000, 2000), 
                                 TxDb = txdb, 
                                 annoDb = params$anno_db, 
                                 verbose = F)
results_sig_anno_df <- results_sig_anno %>% as.data.frame()

plotAnnoPie(results_sig_anno)

plotDistToTSS(results_sig_anno)

anno_data <- toGRanges(txdb, feature = 'gene')
results_sig_anno_batch <- annotatePeakInBatch(results_report_sig, 
                                              AnnotationData = anno_data,
                                              output = 'overlapping',
                                              maxgap = 1000)

results_sig_anno_batch_df <- results_sig_anno_batch %>% as.data.frame()
```