---
title: "Differential Expression"
author: "Harvard Chan Bioinformatics Core"
date: "`r Sys.Date()`"
output:
   html_document:
      code_folding: hide
      df_print: paged
      highlights: pygments
      number_sections: true
      self_contained: true
      theme: default
      toc: true
      toc_float:
         collapsed: true
         smooth_scroll: true
editor_options: 
  chunk_output_type: console
params:
  numerator: tumor
  denominator: normal
  subset_value: NA
  ruv: true
  params_file: ./inst/rmarkdown/templates/rnaseq/skeleton/DE/params_de.R
  project_file: ./inst/rmarkdown/templates/rnaseq/skeleton/information.R
  functions_file: ./inst/rmarkdown/templates/rnaseq/skeleton/load_data.R
  
---

```{r load_params, echo = F}
source(params$params_file)
source(params$project_file)
source(params$functions_file)
```

```{r load_libraries, cache = FALSE, message = FALSE, warning=FALSE, echo=FALSE,}
library(rtracklayer)
library(DESeq2)
library(tidyverse)
library(stringr)
library(RUVSeq)
library(DEGreport)
library(ggpubr)
library(msigdbr)
library(fgsea)
library(org.Hs.eg.db)
library(knitr)
library(EnhancedVolcano)
library(bcbioR)
library(ggprism)
library(viridis)
library(pheatmap)
library(janitor)
colors=cb_friendly_cols(1:15)
ggplot2::theme_set(theme_prism(base_size = 14))
opts_chunk[["set"]](
    cache = F,
    cache.lazy = FALSE,
    dev = c("png", "pdf"),
    error = TRUE,
    highlight = TRUE,
    message = FALSE,
    prompt = FALSE,
    tidy = FALSE,
    warning = FALSE,
    echo = T, 
    fig.height = 4)

# set seed for reproducibility
set.seed(1234567890L)
```

```{r sanitize_datatable}
sanitize_datatable = function(df, ...) {
 # remove dashes which cause wrapping
 DT::datatable(df, ..., rownames=gsub("-", "_", rownames(df)),
                   colnames=gsub("-", "_", colnames(df)))
}
```

# Overview

-   Project: `r project`
-   PI: `r PI`
-   Analyst: `r analyst`
-   Experiment: `r experiment`
-   Aim: `r aim`
-   Comparison: `r paste0(params$subset_value, ': ', params$numerator, ' vs. ', params$denominator)`

```{r create_filenames}

if (!is.na(params$subset_value)){
  filenames = str_interp("${params$subset_value}_${params$numerator}_vs_${params$denominator}")
} else {
  filenames = str_interp("${params$numerator}_vs_${params$denominator}")
}

contrasts = c(column,params$numerator,params$denominator)
coef=paste0(column,"_",params$numerator,"_vs_",params$denominator)

name_expression_fn=file.path(
                             basedir,
                             str_interp("${filenames}_expression.csv"))
name_deg_fn=file.path(
                      basedir,
                      str_interp("${filenames}_deg.csv"))
name_pathways_fn=file.path(
                           basedir,
                           str_interp("${filenames}_pathways.csv"))

```

```{r load_coldata}
coldata <- load_coldata(coldata_fn, params$numerator, params$denominator, subset_column, params$subset_value)
```

```{r load metrics}
metrics <- load_metrics(se_object, multiqc_data_dir) %>% 
  left_join(coldata, by = c('sample' = 'description'))
rownames(metrics) <- metrics$sample
```

```{r load_counts_data}

counts <- load_counts(counts_fn)
counts <- counts[,colnames(counts) %in% coldata$description]

# if the names don't match in order or string check files names and coldata information
counts = counts[rownames(coldata)]
stopifnot(all(names(counts) == rownames(coldata)))

rdata = AnnotationDbi::select(org.Hs.eg.db, rownames(counts), 'SYMBOL', 'ENSEMBL') %>%
  dplyr::select(gene_id = ENSEMBL, gene_name = SYMBOL)

```

```{r setup_RUV}

dds_to_use <- DESeqDataSetFromMatrix(counts, coldata, design = ~1)

vsd_before <- vst(dds_to_use)
norm_matrix = assay(vsd_before)
```

# Covariate analysis

```{r covariates, fig.height = 6, fig.width = 10}
degCovariates(
  norm_matrix,
  metrics %>% as.data.frame %>% column_to_rownames('sample')
)
```

# PCA analysis

```{r before_RUV} 

pca1 <- degPCA(norm_matrix, colData(dds_to_use), 
       condition = column) + ggtitle('PCA')
pca1 + scale_color_cb_friendly()

```

```{r init_DESEQ}
formula <- as.formula(paste0("~ ", " + ", column))
## Check if sample name matches
stopifnot(all(names(counts) == rownames(coldata)))

dds_to_use <- DESeqDataSetFromMatrix(counts, coldata, design = formula)

vsd_before <- vst(dds_to_use)
norm_matrix = assay(vsd_before)
```

```{r, eval=params$ruv, results='asis', echo=FALSE}
cat("# Remove Unwanted Variability

When performing differential expression analysis, it is important to ensure that any detected differences are truly a result of the experimental comparison being made and not any additional variability in the data.")

```

```{r do_RUV, eval=params$ruv}
# If you want to skip the code, just set up formula to be your model in the next chunk of code
design <- coldata[[column]]
diffs <- makeGroups(design)
dat <- norm_matrix
ruvset <- RUVs(dat, cIdx=rownames(dat), k=1, diffs, isLog = T, round = F)
vars <- ruvset$W

new_cdata <- cbind(coldata, vars)

formula <- as.formula(paste0("~ ", 
                          paste0(
                            colnames(new_cdata)[grepl("W", colnames(new_cdata))], 
                            collapse = " + "
                          ), " + ", column)
)
norm_matrix=ruvset$normalizedCounts
pca2 <- degPCA(norm_matrix, new_cdata, 
       condition = column) + ggtitle('After RUV')
pca2 + scale_color_cb_friendly()

```

```{r after_RUV, eval=params$ruv}

dds_to_use <- DESeqDataSetFromMatrix(counts, new_cdata, design = formula)
vsd_to_use<- vst(dds_to_use, blind=FALSE)

```


# Differential Expression

Differential gene expression analysis of count data was performed using the Bioconductor R package, DESeq2, which fits the count data to a negative binomial model. 

Before fitting the model, we often look at a metric called dispersion, which is a measure for variance which also takes into consideration mean expression. A dispersion value is estimated for each individual gene, then 'shrunken' to a more accurate value based on expected variation for a typical gene exhibiting that level of expression. Finally, the shrunken dispersion value is used in the final GLM fit. 

We use the below dispersion plot, which should show an inverse relationship between dispersion and mean expression, to get an idea of whether our data is a good fit for the model. 

```{r DE}
de <- DESeq(dds_to_use)

DESeq2::plotDispEsts(de)
```

Because it is difficult to accurately detect and quantify the expression of lowly expressed genes, differences in their expression between treatment conditions can be unduly exaggerated after the model is fit. We correct for this so that gene LFC is not dependent overall on basal gene expression level.

```{r lfc_shrink}
# resultsNames(de) # check the order is right
resLFC = results(de, contrast=contrasts)
resLFCS <- lfcShrink(de, coef=coef, type="apeglm")

res <- as.data.frame(resLFCS) %>%
  rownames_to_column('gene_id') %>% left_join(rdata, by = 'gene_id') %>% 
  relocate(gene_name) %>% dplyr::rename(lfc = log2FoldChange) %>%
  mutate(pi = abs(lfc) * -log10(padj)) %>% arrange(-pi)

res_sig <- res %>% filter(padj < 0.05) %>% arrange(padj) %>%
  mutate(gene_name = ifelse(is.na(gene_name), gene_id, gene_name))

res_mod <- res %>% mutate(lfc = replace(lfc, lfc < -5, -5)) %>% mutate(lfc = replace(lfc, lfc > 5, 5))
show <- as.data.frame(res_mod[1:10, c("lfc", "padj", "gene_name")])

degMA(as.DEGSet(resLFC)) + ggtitle('Before LFC Shrinking')
```

## MA plot

```{r after_lfc_shrink}
degMA(as.DEGSet(resLFCS), limit = 2) + ggtitle('After LFC Shrinking')

```

## Volcano plot

This volcano plot shows the genes that are significantly up- and down-regulated as a result of the analysis comparison. The points highlighted in red are genes that have padj < 0.05 and a log2-fold change > 1. Points in blue have a padj < 0.05 and a log2-fold change < 1 and points in green have a padj > 0.05 and a log2-fold change > 2. Grey points are non-significant. The dashed lines correspond to the cutoff values of log2 foldchance and padj that we have chosen.

```{r volcano_plot, fig.height=6}
# degVolcano(res_mod[,c('lfc', 'padj')], plot_text = show) 
EnhancedVolcano(res_mod,
                lab= res_mod$gene_name, 
                pCutoff = 1.345719e-03, 
                selectLab = c(res_sig$gene_name[1:15]),
                FCcutoff = 0.5,
                x = 'lfc',
                y = 'padj', 
                title="Volcano Tumor vs. Normal",
                col=as.vector(colors[c("dark_grey", "light_blue",
                                         "purple", "purple")]),
                subtitle = "", xlim=c(-5,5)) 

```

## Heatmap

```{r heapmap}
### Run pheatmap using the metadata data frame for the annotation
ma=norm_matrix[res_sig$gene_id,]
colma=coldata[,c("sample_type"), drop=FALSE]
colors=lapply(colnames(colma), function(c){
  l.col=colors[1:length(unique(colma[[c]]))]
  names(l.col)=unique(colma[[c]])
  l.col
})
names(colors)=colnames(colma)
pheatmap(ma, 
         color = inferno(10), 
         cluster_rows = T, 
         show_rownames = F,
         annotation = colma, 
         annotation_colors = colors,
         border_color = NA, 
         fontsize = 10, 
         scale = "row", 
         fontsize_row = 10, 
         height = 20)
```


## Differentially Expressed Genes

```{r sig_genes_table}
res_sig %>% sanitize_datatable
```

## Plot top 16 genes

```{r top n DEGs, fig.height = 6, fig.width = 8}
n = 16
top_n <- res_sig %>% slice_min(order_by = padj, n = n, with_ties = F) %>% 
  dplyr::select(gene_name, gene_id)
top_n_exp <- norm_matrix %>% as.data.frame() %>% 
  rownames_to_column('gene_id') %>%
  # dplyr::select(-group, -group_name) %>% 
  pivot_longer(!gene_id, names_to = 'sample', values_to = 'log2_expression') %>%
  right_join(top_n, relationship = "many-to-many") %>%
  left_join(coldata, by = c('sample' = 'description'))

ggplot(top_n_exp, aes_string(x = column, y = 'log2_expression')) +
  geom_boxplot(outlier.shape = NA, linewidth=0.5, color="grey") + 
  geom_point() +
  facet_wrap(~gene_name) + 
  ggtitle(str_interp('Expression of Top ${n} DEGs'))

```

# Pathway Enrichment

From the set of differentially expressed genes and using publicly available information about gene sets involved in biological processes and functions, we can calculate which biological processes and functions are significantly perturbed as a result of the treatment. 

```{r}
universe=res %>% 
  filter(!is.na(padj)) %>% pull(gene_id)
mapping = AnnotationDbi::select(org.Hs.eg.db, universe, 'ENTREZID', 'ENSEMBL')

all_in_life=list(
  msigdbr(species = "human", category = "H") %>% mutate(gs_subcat="Hallmark"),
  msigdbr(species = "human", category = "C2", subcategory = "CP:REACTOME"),
  msigdbr(species = "human", category = "C2", subcategory = "CP:KEGG"),
  msigdbr(species = "human", category = "C2", subcategory = "CP:PID"),
  msigdbr(species = "human", category = "C5", subcategory = "GO:BP"),
  msigdbr(species = "human", category = "C5", subcategory = "GO:MF"),
  msigdbr(species = "human", category = "C5", subcategory = "HPO"),
  msigdbr(species = "human", category = "C3", subcategory = "TFT:GTRD"),
  msigdbr(species = "human", category = "C6") %>% mutate(gs_subcat="Oncogenic")
)

ora_input = res %>% filter(!is.na(padj), padj<0.01, abs(lfc)>0.3) %>% pull(gene_id)
input_entrezid <- AnnotationDbi::select(org.Hs.eg.db, ora_input, 'ENSEMBL', columns = c('ENTREZID', 'SYMBOL'))

total_deg=length(unique(ora_input))/length(unique(mapping$ENTREZID))
pathways_ora_all = lapply(all_in_life, function(p){
  pathway = split(x = p$entrez_gene, f = p$gs_name)
  db_name = paste(p$gs_cat[1], p$gs_subcat[1],sep=":")
  respath <- fora(pathways = pathway, 
                  genes = unique(input_entrezid$ENTREZID),
                  universe = unique(mapping$ENTREZID),
                  minSize  = 15,
                  maxSize  = 500)
  coll_respath = collapsePathwaysORA(respath[order(pval)][padj < 0.1], 
                                     pathway, unique(input_entrezid$ENTREZID), unique(mapping$ENTREZID))
  as_tibble(respath[pathway %in% coll_respath$mainPathways])  %>% 
    mutate(database=db_name, NES=(overlap/size)/(total_deg))
}) %>% bind_rows() %>% 
  mutate(analysis="ORA")
  
ora_tb = pathways_ora_all %>% unnest(overlapGenes) %>%
  group_by(pathway) %>% 
  left_join(mapping, by =c("overlapGenes"="ENTREZID")) %>% 
  dplyr::select(pathway, padj, NES, ENSEMBL, analysis,
                database)

pathways_long = ora_tb

```


```{r pathaways_table}
pathways_ora_all %>% sanitize_datatable()
```


```{r write-files}
counts_norm=norm_matrix %>% as.data.frame() %>% 
  rownames_to_column("gene_id") %>% 
  mutate(comparison = str_interp("${params$numerator}_vs_${params$denominator}"))

res_for_writing <- res %>% 
  mutate(comparison = str_interp("${params$numerator}_vs_${params$denominator}"))

pathways_for_writing <- pathways_long %>% 
  mutate(comparison = str_interp("${params$numerator}_vs_${params$denominator}"))
         
if (!is.na(params$subset_value)){
  counts_norm <- counts_norm %>% 
    mutate(subset = params$subset_value)  
  res_for_writing <- res_for_writing %>% 
    mutate(subset = params$subset_value)
  pathways_for_writing <- pathways_for_writing %>% 
    mutate(subset = params$subset_value)
}

write_csv(counts_norm, name_expression_fn)
write_csv(res_for_writing, name_deg_fn)
write_csv(pathways_for_writing, name_pathways_fn)
```
# R session

List and version of tools used for the DE report generation.

```{r}
sessionInfo()
```
