---
title: "Quality Control TEAseq"
title: "Differential Expression"
author: "Harvard Chan Bioinformatics Core"
date: "`r Sys.Date()`"
output:
   html_document:
      code_folding: hide
      df_print: paged
      highlights: pygments
      number_sections: true
      self_contained: true
      theme: default
      toc: true
      toc_float:
         collapsed: true
         smooth_scroll: true
editor_options: 
  chunk_output_type: console
params:
    maxGenes: !r Inf
    minGenes: 750
    maxUMIs: !r Inf
    minUMIs: 1000
    maxMitoPer: 20
    min_novelty: 0.85
    min_cells_per_gene: 10
    min_reads_per_cell: 1000
    max_nCount_ADT: !r Inf
    min_nCount_ADT: 400
    max_nFeature_ADT: 135
    min_nFeature_ADT: 75
    max_PRF: !r Inf 
    min_PRF: 500
    max_FRiP: !r Inf
    min_FRiP: 25
    min_TSS: 2 
    max_NS: 2
    max_blacklistratio: 0.02
    data_dir: !r file.path("data")
    results_dir: !r file.path("results")
output:
    html_document:
        code_folding: hide
        df_print: paged
        highlight: tango
        number_sections: false
        self_contained: true
        theme: paper
        toc: true
        toc_float:
            collapsed: true
            smooth_scroll: false
---

```{r render}
# Set seed for reproducibility
set.seed(1454944673L)
library(knitr)
opts_chunk[["set"]](
    audodep = TRUE,
    cache = FALSE,
    cache.lazy = FALSE,
    error = TRUE,
    fig.height = 6L,
    fig.width = 6L,
    message = FALSE,
    tidy = TRUE,
    warning = FALSE
)
```

```{r, cache = FALSE, message = FALSE, warning=FALSE}
# This set up the working directory to this file so all files can be found
library(rstudioapi)
setwd(fs::path_dir(getSourceEditorContext()$path))
# NOTE: This code will check version, this is our recommendation, it may work
#.      other versions
stopifnot(R.version$major>= 4) # requires R4
stopifnot(compareVersion(R.version$minor,"3.3")==0) # requires >=4.3.3
stopifnot(compareVersion(BiocManager::version(), "3.18")>=0)
```

This code is in this ![](https://img.shields.io/badge/status-draft-grey) revision.

```{r setup, cache=FALSE, message=FALSE}
library(Seurat)
library(tidyverse)
library(ggplot2)
library(ggrepel)
library(cowplot)
library(Matrix)
library(data.table)
library(magrittr)
library(kableExtra)
library(Signac)
library(EnsDb.Hsapiens.v86)
library(qs)
library(bcbioR)
invisible(mapply(
    FUN = dir.create,
    path = c(params$data_dir, params$results_dir),
    MoreArgs = list(showWarnings = FALSE, recursive = TRUE)
))

ggplot2::theme_set(theme_light(base_size = 14))
```


```{r functions, cache=FALSE, message=FALSE}
# Load functions
`%!in%` = Negate(`%in%`)
```

# Overview

```{r}
# Run QC-01-load_data.R and then
# Run QC-02-run_analysis.R
stopifnot(file.exists('data/tea_seurat_unfiltered_clustered.qs'))
seurat <- qread('data/tea_seurat_unfiltered_clustered.qs')

stopifnot(file.exists('data/hashtag_counts.csv'))
hashtag_counts <- read.csv('data/hashtag_counts.csv')
```

- Principal Investigator: 
- Experiment: TEA-Seq of ....

The downloaded data are cellranger-arc (for gene expression and ATAC-Seq) and cellranger (for gene expression and ADT) outputs. Alignment was done using reference genome GRCh38. We use the shared cell barcodes from both results to create a Seurat object that has all 3 modalities.  

First of all, we demultiplex the data so we know which cell belongs to which sample. Below shows the number of cells for each hashtag after demultiplexing. Doublets and Negatives are removed. Overall we have good number of cells per sample. 

```{r}
hashtag_counts <- hashtag_counts %>% plyr::rename(c('Hashtag'='hash.ID'))
# hashtag_counts %>% kable() %>% kable_styling()
```


```{r}
# sample info
sampleinfo = c("Hashtag1: no drug (DMSO)=Control
Hashtag2: Romidepsin=RMD
Hashtag3: Bryostatin=BRY
Hashtag4: AZD=AZD
Hashtag5: RB=RMD+BRY
Hashtag6: RA=RMD+AZD
Hashtag7: PMAi=PMAi")

sampleinfo <- as.data.frame(matrix(unlist(str_split(sampleinfo[1], '\n|: |=')),
                                   ncol = 3, byrow = T)) %>%
  setNames(c('hash.ID', 'Description', 'Sample')) 

sampleinfo %>% full_join(hashtag_counts, by = 'hash.ID') %>% kable() %>% kable_styling()
```

Then we normalize and cluster the data for each modality separately for QC purpose. 

# RNA {.tabset}

Let's first check the gene expression data. 

## Number of cells per sample 


```{r}
# Visualize the number of counts per sample
seurat@meta.data %>%  
  ggplot() + 
  geom_bar(aes(x=sample)) + 
  theme(axis.text.x = element_text(angle = 90, vjust = 1, hjust=1),
        plot.title = element_text(hjust=0.5)) +
  ggtitle("Number of Cells per Sample")
```

## Reads per sample


```{r warning=F, message=F}
seurat@meta.data %>% 
  group_by(sample) %>% 
  summarise(Reads = sum(nCount_RNA)) %>%
  ggplot(aes(x=sample, y=Reads)) + 
  geom_bar(stat = "identity") +
  theme(axis.text.x =element_text(angle = 90,vjust = 1, hjust=1)) +
  ggtitle("Number of Reads per Sample")
```

## UMI counts per cell {.tabset}

Now let's assess the distribution of unique molecular identifier (UMI)-deconvoluted counts per cell. The UMI counts per cell should be generally above 500. The cutoff is at `r params$minUMIs`.

```{r warning=FALSE, message=FALSE, fig.height=8, fig.width=12, results='asis'}
cat('### Histogram \n\n')
seurat@meta.data %>%
  ggplot(aes(x = nCount_RNA)) + 
  geom_histogram(binwidth = 50) + 
  xlab("nUMI") +
  facet_wrap(~sample) + 
  geom_vline(xintercept = params$minUMIs)

cat('\n### Violin plot \n\n')
VlnPlot(seurat, features="nCount_RNA", pt.size = 0) + 
  scale_fill_cb_friendly() +
  geom_hline(yintercept = params$minUMIs)
```

## Genes detected per cell {.tabset}

Here by "detected", we mean genes with a non-zero count measurement per cell. Seeing gene detection in the range of `500`-`5000` is normal for most single-cell experiments.
The cutoff is at `r params$minGenes`.

```{r warning=FALSE, message=FALSE, fig.height=8, fig.width=12, results='asis'}
cat('### Histogram \n\n')
seurat@meta.data %>% 
  ggplot(aes(x = nFeature_RNA)) + 
  geom_histogram(binwidth = 50) + 
  xlab("nGene") +
  facet_wrap(~sample) +
  geom_vline(xintercept = params$minGenes)

cat('\n### Violin plot \n\n')
VlnPlot(seurat, features="nFeature_RNA", pt.size = 0) +
    scale_fill_cb_friendly() +
  geom_hline(yintercept = params$minGenes)
```


## Mitochondrial abundance {.tabset}

We evaluate overall mitochondrial gene expression as a biomarker of cellular stress during sample preparation. The mitochondrial percentages are mostly below 20%.


```{r warning=F, message=F, fig.height=8, fig.width=12, results='asis'}
cat('### Histogram \n\n')
seurat@meta.data %>% 
  ggplot(aes(x = percent_mito)) + 
  geom_histogram() + 
  facet_wrap(~sample) +
  geom_vline(xintercept = params$maxMitoPer)

cat('\n### Violin plot \n\n')
VlnPlot(seurat, features="percent_mito", pt.size = 0) +
  scale_fill_cb_friendly() +
  geom_hline(yintercept = params$maxMitoPer)
```

## Novelty {.tabset}

Another way to QC the data is to look for less novelty, that is cells that have less genes detected per count than other cells. We can see the samples where we sequenced each cell less have a higher overall novelty, that is because we have not started saturated the sequencing for any given gene for these samples. Outlier cells in these samples might be cells that we have a less complex RNA species than other cells. Sometimes we can detect contamination with low complexity cell types like red blood cells via this metric.

All cells have high novelty scores (> 0.85). 

```{r novelty, fig.height=8, fig.width=12, results='asis'}
cat('### Histogram \n\n')
# Visualize the overall novelty of the gene expression by visualizing the genes detected per UMI
seurat@meta.data %>%
  ggplot(aes(x=novelty)) +
  geom_histogram(bins=50) +
  facet_wrap(~sample) 

cat('\n### Violin plot \n\n')
VlnPlot(seurat, features="novelty", pt.size = 0) +
    scale_fill_cb_friendly()
```

## UMIs vs. genes detected {.tabset}

If we graph out the total number of UMI counts per cell vs. the genes detected per cell, we can assess whether there is a large population of low quality cells with low counts and/or gene detection. For some of the samples there is a cluster of cells that have increasing UMIs yet gene detection is low. We will want to filter out these low complexity cells. The data points are colored by different metrics to see if there is some correlation.

### Data points colored by Percent mitochondrial genes

```{r umis_vs_genes_mito, warning=F, message=F, fig.height=8, fig.width=12}
# Visualize the correlation between genes detected and number of UMIs and determine whether strong presence of cells with low numbers of genes/UMIs
seurat@meta.data %>%
  ggplot(aes(x=nCount_RNA, y=nFeature_RNA, color=percent_mito)) + 
  geom_point() + 
  stat_smooth(method=lm) +
  scale_x_log10() + 
  scale_y_log10() + 
  geom_vline(xintercept = params$minUMIs) +
  geom_hline(yintercept = params$minGenes) +
  xlab("nUMI") + ylab("nGene") +
  facet_wrap(~sample)
```

### Data points colored by Novelty

```{r umis_vs_genes_novelty, warning=F, message=F, fig.height=8, fig.width=12}
# Visualize the correlation between genes detected and number of UMIs and determine whether strong presence of cells with low numbers of genes/UMIs
seurat@meta.data %>% 
  ggplot(aes(x=nCount_RNA, y=nFeature_RNA, color=novelty)) + 
  geom_point() + 
  stat_smooth(method=lm) +
  scale_x_log10() + 
  scale_y_log10() + 
  geom_vline(xintercept = params$minUMIs) +
  geom_hline(yintercept = params$minGenes) +
  xlab("nUMI") + ylab("nGene") +
  facet_wrap(~sample)
```


## UMAP plots {.tabset}

### Color by sample

Clustering of the cells is similar to the first dataset.

```{r results='asis'}
DimPlot(object = seurat, reduction = "umap", group.by = "sample", order = T) +
    scale_color_cb_friendly()

cat('\n\n')
```

```{r results='asis', fig.width=12}
DimPlot(object = seurat, reduction = "umap", split.by = 'sample')+
    scale_color_cb_friendly()
cat('\n\n')
```

### Color by cell cycle

```{r results='asis'}
DimPlot(object = seurat, reduction = "umap", group.by = "Phase", order = T) +
    scale_color_cb_friendly()
cat('\n\n')
```

```{r results='asis', fig.width=12}
DimPlot(object = seurat, reduction = "umap", group.by = "Phase", split.by = 'sample') +
      scale_color_cb_friendly()
cat('\n\n')
```

### Color by UMI counts

```{r results='asis', fig.width=12}
FeaturePlot(seurat, reduction = "umap", features = "nCount_RNA", split.by = "sample", cols = c("light blue", "red")) & theme(legend.position = c(0.2,0.2))
cat('\n\n')
```
### Color by the number of detected genes

```{r results='asis', fig.width=12}
FeaturePlot(seurat, reduction = "umap", features = "nFeature_RNA", split.by = "sample", cols = c("light blue", "red"))  & theme(legend.position = c(0.2,0.2))
cat('\n\n')
```
### Color by mito percentage

```{r results='asis', fig.width=12}
FeaturePlot(seurat, reduction = "umap", features = "percent_mito", split.by = "sample", cols = c("light blue", "red")) & theme(legend.position = c(0.2,0.2))
cat('\n')
```
### Color by novelty

```{r results='asis', fig.width=12}
FeaturePlot(seurat, reduction = "umap", features = "novelty", split.by = "sample", cols = c("light blue", "red")) & theme(legend.position = c(0.2,0.2))
cat('\n')
```

# ADT {.tabset}

The numbers of detected proteins are more than the the first experiment. The cells are clustered slightly differently from gene expression and ATAC-Seq results on the UMAP and the first dataset. 

## UMI counts per cell

```{r warning=FALSE, message=FALSE, results='asis'}
VlnPlot(seurat, features = "nCount_ADT", ncol = 1, pt.size = 0) +
  scale_fill_cb_friendly() +
  xlab("") +
  ylab("UMI")
cat('\n\n')
```

**Zoom in** (cutoff line is at `r params$min_nCount_ADT`):

```{r warning=FALSE, message=FALSE, results='asis'}
VlnPlot(seurat, features = "nCount_ADT", ncol = 1, pt.size = 0) +
  scale_fill_cb_friendly() +
  xlab("") +
  ylab("UMI") +
  ylim(0,3000) +  
  geom_hline(yintercept = params$min_nCount_ADT)
cat('\n\n')
```

## Detected proteins per cell

Lines are drawn at `r params$min_nFeature_ADT` and `r params$max_nFeature_ADT`.

```{r warning=FALSE, message=FALSE, results='asis'}
VlnPlot(seurat, features = "nFeature_ADT", ncol = 1, pt.size = 0) +
  scale_fill_cb_friendly()+
  xlab("") +
  geom_hline(yintercept = params$max_nFeature_ADT) +  
  geom_hline(yintercept = params$min_nFeature_ADT) +
  ylab("Number of detected proteins")
cat('\n\n')
```

## UMAP plots {.tabset}

### Color by sample

We can still see cells are clustered by sample, but not as clear as the first dataset.

```{r results='asis'}
DimPlot(object = seurat, reduction = "umapADT", group.by = "sample", order = T) +
      scale_color_cb_friendly()
cat('\n\n')
```
```{r results='asis', fig.width=12}
DimPlot(object = seurat, reduction = "umapADT", split.by = 'sample') +
      scale_color_cb_friendly()
cat('\n\n')
```


### Color by UMI counts

```{r results='asis', fig.width=12}
FeaturePlot(seurat, reduction = "umapADT", features = "nCount_ADT", split.by = "sample", cols = c("light blue", "red")) & theme(legend.position = c(0.2,0.2))
cat('\n\n')
```
### Color by the number of detected proteins

```{r results='asis', fig.width=12}
FeaturePlot(seurat, reduction = "umapADT", features = "nFeature_ADT", split.by = "sample", cols = c("light blue", "red"))  & theme(legend.position = c(0.2,0.2))
cat('\n\n')
```



# ATAC-Seq {.tabset}

We use some of the results from cellranger outputs and the peaks called using MACS2 for QCing the scATAC-Seq data. 

## UMI counts per cell

```{r warning=FALSE, message=FALSE, results='asis'}
VlnPlot(seurat, features = "nCount_MACS2peaks", ncol = 1, pt.size = 0) +
  scale_fill_cb_friendly() +
  xlab("") +
  ylab("UMI")

cat('\n\n')
```
## Detected peaks per cell

```{r warning=FALSE, message=FALSE}
VlnPlot(seurat, features = "nFeature_MACS2peaks", ncol = 1, pt.size = 0) +
  scale_fill_cb_friendly() +
  xlab("") +
  ylab("UMI")

VlnPlot(seurat, features = "nFeature_CellRangerPeaks", ncol = 1, pt.size = 0) +
  scale_fill_cb_friendly() +
  xlab("") +
  ylab("UMI")
```

## QC metrics {.tabset}

### Total number of fragments in peaks {.tabset}

This metric represents the total number of fragments (= reads) mapping within a region of the genome that is predicted to be accessible (= a peak). It's a measure of cellular sequencing depth / complexity. Cells with very few reads may need to be excluded due to low sequencing depth. Cells with extremely high levels may represent doublets, nuclei clumps, or other artefacts. 


```{r results='asis', warning=FALSE, message=FALSE}
DefaultAssay(seurat) <- "MACS2peaks"

cat('#### Histogram \n\n')
seurat@meta.data %>% 
  ggplot(aes(x = atac_peak_region_fragments)) + 
  geom_histogram() + 
  facet_wrap(~sample) +
  geom_vline(xintercept = params$min_PRF)
 

cat('\n#### Violin plot\n\n')
VlnPlot(
  object = seurat,
  features = 'atac_peak_region_fragments',
  pt.size = 0
) + ylab('Total number of fragments in peaks') + 
  NoLegend() + 
  geom_hline(yintercept = params$min_PRF) + xlab("") +
  scale_fill_cb_friendly()

cat('\n\n')
```

### Fraction of fragments in peaks

It represents the fraction of all fragments that fall within ATAC-seq peaks. Cells with low values (i.e. <15-20%) often represent low-quality cells or technical artifacts that should be removed. Note that this value can be sensitive to the set of peaks used.

```{r results='asis'}
VlnPlot(
  object = seurat,
  features = 'pct_reads_in_peaks',
  pt.size = 0
) + NoLegend() + xlab("") +
  scale_fill_cb_friendly() +
  geom_hline(yintercept = params$min_FRiP) 

cat('\n\n')
```


### Transcriptional start site (TSS) enrichment score {.tabset}

The ENCODE project has defined an ATAC-seq targeting score based on the ratio of fragments centered at the TSS to fragments in TSS-flanking regions (see https://www.encodeproject.org/data-standards/terms/). Poor ATAC-seq experiments typically will have a low TSS enrichment score. We can compute this metric for each cell with the TSSEnrichment() function, and the results are stored in metadata under the column name TSS.enrichment.


```{r results='asis'}
VlnPlot(
  object = seurat,
  features = 'TSS.enrichment',
  pt.size = 0
) +   scale_fill_cb_friendly() +
NoLegend() + xlab("")
cat('\n\n')
```
The following tabs show the TSS enrichment score distribution for each sample. Cells with high-quality ATAC-seq data should show a clear peak in reads at the TSS, with a decay the further we get from it.

Each plot is split between cells with a high or low global TSS enrichment score (cuffoff at `r params$min_TSS`), to double-check whether cells with lowest enrichment scores still follow the expected pattern or rather need to be excluded.

```{r results='asis'}
seurat$TSS.group <- ifelse(seurat$TSS.enrichment > params$min_TSS, "High", "Low")

for (sample in sort(unique(seurat$sample))){
  cat('####', sample, '\n\n')
  p <- TSSPlot(subset(x = seurat, idents = sample), group.by = "TSS.group") + NoLegend()
  print(p)
  cat('\n\n')
}
```

### Nucleosome signal

The histogram of DNA fragment sizes (determined from the paired-end sequencing reads) should exhibit a strong nucleosome banding pattern corresponding to the length of DNA wrapped around a single nucleosome, i.e peaks at approximately 100bp (nucleosome-free), and mono, di and tri nucleosome-bound peaks at 200, 400 and 600bp. We calculate this per single cell, and quantify the approximate ratio of mononucleosomal to nucleosome-free fragments (stored as nucleosome_signal). Cells with lower nucleosome signal have a higher ratio of nucleosome-free fragments.

```{r warning = FALSE}
seurat$nucleosome_group <- ifelse(seurat$nucleosome_signal > 1, "high NS", "low NS")
FragmentHistogram(seurat, group.by = 'nucleosome_group', cells = colnames(seurat[, seurat$sample == 'Control']))
```

```{r }
VlnPlot(
  object = seurat,
  features = 'nucleosome_signal',
  pt.size = 0
) +   scale_fill_cb_friendly() +
NoLegend() + xlab("")
```




### Blacklist ratio

tIt's he ratio of reads in genomic blacklist regions. The [ENCODE project](https://www.encodeproject.org/) has provided a list of [blacklist regions](https://github.com/Boyle-Lab/Blacklist), representing reads which are often associated with artefactual signal. Cells with a high proportion of reads mapping to these areas (compared to reads mapping to peaks) often represent technical artifacts and should be removed. ENCODE blacklist regions for human (hg19 and GRCh38), mouse (mm10), Drosophila (dm3), and C. elegans (ce10) are included in the Signac package. **Peaks overlapping with the balcklist regions were removed in the analysis, so we don't show blacklist fraction here**.

Line is drawn at `r params$min_blacklistratio`. 

```{r}
VlnPlot(
  object = seurat,
  features = 'blacklist_fraction',
  pt.size = 0
) +   scale_fill_cb_friendly() +
  NoLegend() +
  geom_hline(yintercept = params$min_blacklistratio) 
  
```


## Normalization, linear dimensional reduction, and clustering

* Normalization: Signac performs term frequency-inverse document frequency (TF-IDF) normalization. This is a two-step normalization procedure, that both normalizes across cells to correct for differences in cellular sequencing depth, and across peaks to give higher values to more rare peaks.

* Feature selection: The low dynamic range of scATAC-seq data makes it challenging to perform variable feature selection, as we do for scRNA-seq. Instead, we can choose to use only the top n% of features (peaks) for dimensional reduction, or remove features present in less than n cells with the FindTopFeatures() function. Here, we will use all features, though we note that we see very similar results when using only a subset of features (try setting min.cutoff to ‘q75’ to use the top 25% all peaks), with faster runtimes. Features used for dimensional reduction are automatically set as VariableFeatures() for the Seurat object by this function.

* Dimension reduction: We next run singular value decomposition (SVD) on the TD-IDF matrix, using the features (peaks) selected above. This returns a reduced dimension representation of the object (for users who are more familiar with scRNA-seq, you can think of this as analogous to the output of PCA).

The combined steps of TF-IDF followed by SVD are known as latent semantic indexing (LSI), and were first introduced for the analysis of scATAC-seq data by [Cusanovich et al. 2015.](https://www.science.org/doi/10.1126/science.aax6234)

The first LSI component often captures sequencing depth (techni ccal variation) rather than biological variation. If this is the case, the component should be removed from downstream analysis. We can assess the correlation between each LSI component and sequencing depth using the DepthCor() function (see below). 
Here we see there is a very strong correlation between the first LSI component and the total number of counts for the cell, so we will perform downstream steps without this component.

```{r results='asis'}
DepthCor(seurat, assay = 'MACS2peaks')
cat('\n\n')
```

## UMAP plots

```{r results='asis'}
DimPlot(object = seurat, group.by = 'sample', reduction = "umapATAC") +
    scale_color_cb_friendly()

cat('\n\n')
```

<!-- # {.unlisted .unnumbered} -->


# Conclusion

We will apply the following filters to **keep the high-quality cells**. For most of the criteria below, we don't have an upper limit, as we have removed doublets after demultiplexing.

**RNA**

* Number of UMIs (nCount_RNA) > `r params$minUMIs`
* Number of detected genes (nFeature_RNA) > `r params$minGenes`
* mitochondrial percentages (nFeature_RNA) < `r params$maxMitoPer`

We don't apply any filter for complexity, as all the cells have relatively high complexity (> 0.85).

**ADT**

* Number of UMIs (nCount_ADT) > `r params$min_nCount_ADT`
* Number of detected proteins (nFeature_ADT) < `r params$max_nFeature_ADT` and > `r params$min_nFeature_ADT`

**ATAC**

* Total number of fragments in peaks (atac_peak_region_fragments) > `r params$min_PRF`
* Percentage of fragments in peaks (pct_reads_in_peaks) > `r params$min_FRiP`
* TSS enrichment (TSS.enrichment) > `r params$min_TSS`
* Nucleosome signal (nucleosome_signal) < `r params$max_NS`
* Blacklisted regions ratio (blacklist_fraction) < `r params$max_blacklistratio`


Number of cells before filtering:

```{r}
table(seurat$sample) %>% as.data.frame() %>% set_colnames(c('sample', '#Cells')) %>% kable() %>% kable_styling()
```


```{r}
# Please select your cutoffs
rna_keep <- seurat@meta.data[(seurat@meta.data$nCount_RNA > params$minUMIs &
                                seurat@meta.data$nFeature_RNA > params$minGenes &
                                seurat@meta.data$percent_mito < params$maxMitoPer), ]
rna_filt <- seurat@meta.data[rownames(seurat@meta.data) %!in% rownames(rna_keep),]

adt_keep <- seurat@meta.data[(seurat@meta.data$nCount_ADT > params$min_nCount_ADT &
                                seurat@meta.data$nFeature_ADT > params$min_nFeature_ADT &
                                seurat@meta.data$nFeature_ADT < params$max_nFeature_ADT), ]

adt_filt <- seurat@meta.data[rownames(seurat@meta.data) %!in% rownames(adt_keep),]

atac_keep <- seurat@meta.data[(seurat@meta.data$atac_peak_region_fragments > params$min_PRF &
                                 seurat@meta.data$pct_reads_in_peaks > params$min_FRiP &
                                 seurat@meta.data$TSS.enrichment > params$min_TSS &
                                 seurat@meta.data$nucleosome_signal < params$max_NS &
                                 seurat@meta.data$blacklist_fraction < params$max_blacklistratio), ]

atac_filt <- seurat@meta.data[rownames(seurat@meta.data) %!in% rownames(atac_keep),]

library(ggvenn)
x = list(
  RNA = rownames(rna_filt),
  ADT = rownames(adt_filt),
  ATAC = rownames(atac_filt)
)
ggvenn(x, names(x))
```

```{r eval=FALSE}
# this is only run once before rendering
unfiltered_seurat <- seurat
seurat <- subset(
  x = seurat,
  subset = nCount_RNA > params$minUMIs &
    nFeature_RNA > params$minGenes &
    percent_mito < params$maxMitoPer &
    nCount_ADT > params$min_nCount_ADT &
    nFeature_ADT > params$min_nFeature_ADT &
    nFeature_ADT < params$max_nFeature_ADT &
    atac_peak_region_fragments > params$min_PRF &
    TSS.enrichment > params$min_TSS &
    nucleosome_signal < params$max_NS &
    blacklist_fraction < params$max_blacklistratio
  )

# reclustering
# RNA
DefaultAssay(seurat) <- 'RNA'
seurat <- SCTransform(seurat, conserve.memory=TRUE)
seurat <- RunPCA(object = seurat, verbose = T)
co <- 40
seurat <- FindNeighbors(object = seurat, reduction = "pca", dims = 1:co)
seurat <- RunUMAP(object = seurat, dims=1:co )

# clustering by ADT data
seurat <- NormalizeData(seurat, assay = "ADT", normalization.method = "CLR", margin = 2)
seurat <- ScaleData(seurat, assay = "ADT")
seurat <- RunPCA(seurat, features = rownames(seurat@assays$ADT@counts), verbose = TRUE, assay= "ADT", reduction.name = "pcaADT", reduction.key = "pcaADT_")
seurat <- RunUMAP(seurat, dims = 1:20, reduction = "pcaADT", assay = "ADT", reduction.name = "umapADT", reduction.key = "umapADT_")

# ATAC data normalization, dimension reduction, and clustering
seurat <- RunTFIDF(seurat, assay = "MACS2peaks")
seurat <- FindTopFeatures(seurat, min.cutoff = 'q0', assay = "MACS2peaks")
seurat <- RunSVD(seurat, assay = "MACS2peaks")
seurat <- RunUMAP(object = seurat, dims = 2:30, reduction = 'lsi', assay = "MACS2peaks", reduction.name = "umapATAC", reduction.key = "umapATAC_" )

qsave(seurat, 'data/tea_seurat_filtered_clustered.qs')
```

Number of cells **after filtering**:

```{r}
# get the filtered seurat
stopifnot(file.exists('data/tea_seurat_filtered_clustered.qs'))
seurat <- qread('data/tea_seurat_filtered_clustered.qs')
table(seurat$sample) %>% as.data.frame() %>% set_colnames(c('sample', '#Cells')) %>% kable() %>% kable_styling()
```

**We take a look at the UMAPs after filtering out low quality cells for each modality.**

```{r eval=FALSE, fig.width=12, fig.height=12}
plot_grid(
  DimPlot(object = seurat, group.by = "sample", reduction = "umap") + ggtitle('RNA'),
  DimPlot(object = seurat, group.by = "sample", reduction = "umapADT") + ggtitle('ADT'),
  DimPlot(object = seurat, group.by = "sample", reduction = "umapATAC") + ggtitle('ATAC')
)
```


# Session info

```{r}
devtools::session_info()
```

