---
title: "Differential Expression"
author: "Harvard Chan Bioinformatics Core"
date: "`r Sys.Date()`"
output:
   html_document:
      code_folding: hide
      df_print: paged
      highlights: pygments
      number_sections: true
      self_contained: true
      theme: default
      toc: true
      toc_float:
         collapsed: true
         smooth_scroll: true
editor_options: 
  chunk_output_type: console
params:
  # Put hg38, mm10, mm39, or other
  
  ## Combatseq and ruv can both be false or ONLY ONE can be true
  ## Both cannot be true
  numerator: tumor
  denominator: normal
  column: sample_type
  subset_column: null
  subset_value: null
  genome: hg38
  ruv: false
  combatseq: false
  params_file: params_de-example.R
  project_file: ../information.R
  functions_file: load_data.R
---


```{r}
# This set up the working directory to this file so all files can be found
library(rstudioapi)
setwd(fs::path_dir(getSourceEditorContext()$path))
```


```{r load_params, cache = FALSE, message = FALSE, warning=FALSE}
# 1. Set up input files in this R file (params_de.R)
source(params$params_file)
# 2. Set up project file (already done from QC probably)
source(params$project_file)
# 3. Load custom functions to load data from coldata/metrics/counts
source(params$functions_file)
# IMPORTANT set these values if you are not using the parameters in the header (lines 22-31)
genome=params$genome
column=params$column
numerator=params$numerator
denominator=params$denominator
subset_column=params$subset_column
subset_value=params$subset_value
run_ruv=params$ruv
run_combatseq=params$combatseq
factor_of_interest <- column
```


```{r load_libraries, cache = FALSE, message = FALSE, warning=FALSE}
library(rtracklayer)
library(DESeq2)
library(tidyverse)
library(stringr)
library(DEGreport)
library(ggpubr)
library(msigdbr)
library(fgsea)
library(org.Hs.eg.db)
library(knitr)
library(EnhancedVolcano)
library(bcbioR)
library(ggprism)
library(viridis)
library(pheatmap)
library(janitor)
library(ggforce)
library(vegan)

colors=cb_friendly_cols(1:15)
ggplot2::theme_set(theme_prism(base_size = 14))
opts_chunk[["set"]](
    cache = F,
    cache.lazy = FALSE,
    dev = c("png", "pdf"),
    error = TRUE,
    highlight = TRUE,
    message = FALSE,
    prompt = FALSE,
    tidy = FALSE,
    warning = FALSE,
    echo = T, 
    fig.height = 4)

# set seed for reproducibility
set.seed(1234567890L)
```

```{r sanitize_datatable}
sanitize_datatable = function(df, ...) {
 # remove dashes which cause wrapping
 DT::datatable(df, ..., rownames=gsub("-", "_", rownames(df)),
                   colnames=gsub("-", "_", colnames(df)))
}
```



```{r load_data, message=F, warning=F}
# This code will load from bcbio or nf-core folder
# NOTE make sure to set numerator and denominator
coldata <- load_coldata(coldata_fn, column,
                        numerator, denominator,
                        subset_column, subset_value)
coldata$sample=row.names(coldata)

counts <- load_counts(counts_fn)
counts <- counts[,colnames(counts) %in% coldata$sample]

metrics <- load_metrics(se_object, multiqc_data_dir, gtf_fn, counts) %>% 
  left_join(coldata, by = c('sample')) %>% 
  as.data.frame()
rownames(metrics) <- metrics$sample
# if the names don't match in order or string check files names and coldata information
counts = counts[,rownames(metrics)]
coldata = coldata[rownames(metrics),]
stopifnot(all(names(counts) == rownames(metrics)))
```



# Overview

-   Project: `r project`
-   PI: `r PI`
-   Analyst: `r analyst`
-   Experiment: `r experiment`
-   Aim: `r aim`
-   Comparison: `r ifelse(is.null(subset_value), paste0(numerator, ' vs. ', denominator), paste0(subset_value, ': ', numerator, ' vs. ', denominator))`

```{r create_filenames}

if (!is.null(subset_value) & !is.null(subset_value)){
  filenames = str_interp("${subset_value}_${numerator}_vs_${denominator}")
} else {
  filenames = str_interp("${numerator}_vs_${denominator}")
}

contrasts = c(column,numerator,denominator)
coef=paste0(column,"_",numerator,"_vs_",denominator)

name_expression_fn=file.path(
                             basedir,
                             str_interp("${filenames}_expression.csv"))
name_deg_fn=file.path(
                      basedir,
                      str_interp("${filenames}_deg.csv"))
name_pathways_fn=file.path(
                           basedir,
                           str_interp("${filenames}_pathways.csv"))

```

```{r load_counts_data}

rdata = AnnotationDbi::select(org.Hs.eg.db, rownames(counts), 'SYMBOL', 'ENSEMBL') %>%
  dplyr::select(gene_id = ENSEMBL, gene_name = SYMBOL) %>% 
  distinct(gene_id, .keep_all = TRUE)

```

```{r setup_RUV}
dds_to_use <- DESeqDataSetFromMatrix(counts, coldata, design = ~1)

vsd_before <- vst(dds_to_use)
norm_matrix = assay(vsd_before)
```


# PCA and group level variance.

**Principal Component Analysis (PCA) is a statistical technique used to simplify high-dimensional data by identifying patterns and reducing the number of variables. In the context of gene expression, PCA helps analyze large datasets containing information about the expression levels of thousands of genes across different samples (e.g., tissues, cells).**

Dispersion estimates are a key part of the DESEQ2 analysis. DESEQ2 uses data from all samples and all genes to generate a relationship between level expression and variance and then shrinks per gene dispersions to match this distribution. If one group has higher variance than all others this will affect the dispersion estimates. Here we visually check that the variance per group is similar using a PCA. The ellipses are minimal volume enclosing ellipses using the Khachiyan algorithm.

**It is best practice NOT to subset your data unless one group has significantly higher variance than the others. The best dispersion estimates are obtained with more data.**

**This code automatically uses the column value from the header. You can also manually add a factor of interest to define the groups. One can be created by combining multiple metadata columns using the paste0 function.**

```{r set group, eval=FALSE, echo=FALSE}
## Example of creating a group covariate

meta$group <- paste0(meta$sex,"_", meta$age,"_",meta$treatment)

factor_of_interest <- "insert column name for covariate of interest"
```


```{r PCA}
pca <- degPCA(norm_matrix, metrics,
              condition = factor_of_interest, name = "sample", data = T)

pca$plot + ggtitle(paste0("All samples", "\nPCA using ", nrow(vsd_before), " genes")) +
  theme(plot.title=element_text(hjust=0.5)) +
  geom_mark_ellipse(aes(color = sample_type)) +  scale_color_cb_friendly()
```

## PERMDISP

Groups in a univariate analysis can also differ with regard to their mean values, variation around those means, or both. In univariate analyses, dispersion can be examined using Levene’s test.  PERMDISP is a multivariate extension of Levene’s test to examine whether groups differ in variability. In essence, PERMDISP involves calculating the distance from each data point to its group centroid and then testing whether those distances differ among the groups. [Source](https://uw.pressbooks.pub/appliedmultivariatestatistics/chapter/permdisp/)

Here we apply this test to our variance stabilized data. We calculate distances between samples and then use the `betadisper()` function from the popular vegan package. We get two overall p-values where significant means that the dispersions are different between groups. The first p-value comes from the `anova()` function and the second from the `permutest()` function. We also get pairwise p-values for every group-group comparison.

```{r PERMDISP}
vare.disa <- vegdist(t(assay(vsd_before)))

mod = betadisper(vare.disa, metrics[[factor_of_interest]])
anova(mod)
permutest(mod, pairwise = TRUE)

```



# Covariate analysis

Multiple factors related to the experimental design or quality of sequencing may influence the outcomes of a given RNA-seq experiment. To further determine whether any confounding covariate risks affecting the results of our differential expression analyses, it is useful to assess the correlation between covariates and principal component (PC) values.

Here, we are using `DEGreport::degCovariates()` to explore potential correlations between variables provided in the metadata and all PCs that account for at least 5% of the variability in the data. If applicable, significant correlations (FDR < 0.1) are circled. **This diagnostic plot helps us determine which variables we may need to add to our DE model.**


```{r covariates, fig.height = 6, fig.width = 10}
degCovariates(
  norm_matrix,
  metrics,
)
```



```{r init_DESEQ}
formula <- as.formula(paste0("~ ", " + ", column))
## Check if sample name matches
stopifnot(all(names(counts) == rownames(coldata)))

dds_to_use <- DESeqDataSetFromMatrix(counts, coldata, design = formula)

vsd_before <- vst(dds_to_use)
norm_matrix = assay(vsd_before)
new_cdata <- coldata
```


```{r, eval=F, echo=FALSE}
#### IF YOU ARE RUNNING RUV OR COMBATSEQ RUN THE CHUNKS BELOW OTHERWISE SKIP TO Differential Expression SECTION

### RUV - LINES 261-296
### COMBATSEQ - LINES 303-369
```



```{r, eval=run_ruv, results='asis', echo=run_ruv}
cat("# Remove Unwanted Variability

When performing differential expression analysis, it is important to ensure that any detected differences are truly a result of the experimental comparison being made and not any additional variability in the data.")
```

```{r do_RUV, eval=run_ruv, echo=run_ruv}
library(RUVSeq)

# If you want to skip the code, just set up formula to be your model in the next chunk of code
design <- coldata[[column]]
diffs <- makeGroups(design)
dat <- norm_matrix
# by default is running one variable, 
# change K parameter to other number to find more unknown covariates
ruvset <- RUVs(dat, cIdx=rownames(dat), k=1, diffs, isLog = T, round = F)
vars <- ruvset$W

new_cdata <- cbind(coldata, vars)

formula <- as.formula(paste0("~ ", 
                          paste0(
                            colnames(new_cdata)[grepl("W", colnames(new_cdata))], 
                            collapse = " + "
                          ), " + ", column)
)
norm_matrix=ruvset$normalizedCounts
pca2 <- degPCA(norm_matrix, new_cdata, 
       condition = column) + ggtitle('After RUV')
pca2 + scale_color_cb_friendly()

```

```{r after_RUV, eval=run_ruv}

dds_to_use <- DESeqDataSetFromMatrix(counts, new_cdata, design = formula)
vsd_to_use<- vst(dds_to_use, blind=FALSE)

```

```{r combat-text , eval=run_combatseq, results='asis', echo=run_combatseq}
library(sva)

cat("# Remove Batch Effects

Here we apply Combat-seq (https://github.com/zhangyuqing/ComBat-seq) to try to remove batch effects so we can better tease out the effects of interest. 

Combat-seq uses a negative binomial regression to model batch effects, providing adjusted data by mapping the original data to an expected distribution if there were no batch effects. The adjusted data preserves the integer nature of counts, so that it is compatible with the assumptions of state-of-the-art differential expression software (e.g. edgeR, DESeq2, which specifically request untransformed count data).")

```


```{r set_variable_combatseq, eval=run_combatseq, echo=run_combatseq}

## FILL OUT THIS CHUNK OF CODE IF YOU WANT TO RUN COMBATSEQ

## Set your batch effect variable here this is the variable that combatseq will try to remove

## Column name of your batch variable
to_remove = "batch"

## Column name of of your variable(s) of interest

to_keep = "sample_type"


coldata[[to_remove]] <- as.factor(coldata[[to_remove]])
coldata[[to_keep]] <- as.factor(coldata[[to_keep]])


batch = coldata[[to_remove]]
treatment = coldata[[to_keep]]

## If you have multiple variables of interest you will need to cbind them into one variable

#treatment1 = metrics[[to_keep]]
#treatment2 = metrics[[to_keep]]
#treatment3 = metrics[[to_keep]]


# imp = cbind(as.numeric(as.character(treatment1)),as.numeric(as.character(treatment2)), as.numeric(as.character(treatment3)))

```


```{r do_combatseq, eval=run_combatseq}
adjusted_counts <- ComBat_seq(as.matrix(counts), batch=batch, group = treatment)

## For multiple variables of interest

# adjusted_counts <- ComBat_seq(as.matrix(counts2), batch=batch, covar_mod = imp)

```

```{r after_combatseq, eval=run_combatseq}
# NOTE: Make sure the formula doens't contain the covariates used in combatseq above
dds_to_use <- DESeqDataSetFromMatrix(adjusted_counts, coldata, design = formula)
vsd_combat<- vst(dds_to_use, blind=FALSE)

combat_matrix = assay(vsd_combat)

pca_combat <- degPCA(combat_matrix, coldata, 
       condition = column) + ggtitle('After Combatseq')
pca_combat + scale_color_cb_friendly()

```


# Differential Expression

Differential gene expression analysis of count data was performed using the Bioconductor R package, DESeq2, which fits the count data to a negative binomial model. 

Before fitting the model, we often look at a metric called dispersion, which is a measure for variance which also takes into consideration mean expression. A dispersion value is estimated for each individual gene, then 'shrunken' to a more accurate value based on expected variation for a typical gene exhibiting that level of expression. Finally, the shrunken dispersion value is used in the final GLM fit. 

We use the below dispersion plot, which should show an inverse relationship between dispersion and mean expression, to get an idea of whether our data is a good fit for the model. 

```{r DE}
de <- DESeq(dds_to_use)

DESeq2::plotDispEsts(de)
```

Because it is difficult to accurately detect and quantify the expression of lowly expressed genes, differences in their expression between treatment conditions can be unduly exaggerated after the model is fit. We correct for this so that gene LFC is not dependent overall on basal gene expression level.

```{r lfc_shrink}
# resultsNames(de) # check the order is right
resLFC = results(de, contrast=contrasts)
resLFCS <- lfcShrink(de, coef=coef, type="apeglm")

res <- as.data.frame(resLFCS) %>%
  rownames_to_column('gene_id') %>% left_join(rdata, by = 'gene_id') %>% 
  relocate(gene_name) %>% dplyr::rename(lfc = log2FoldChange) %>%
  mutate(pi = abs(lfc) * -log10(padj)) %>% arrange(-pi)

## Filter out genes that have no expression or were filtered out by DESEQ2
res <- res[res$baseMean>0,] %>% drop_na(padj) %>% drop_na(pvalue)

res_sig <- res %>% filter(padj < 0.05) %>% arrange(padj) %>%
  mutate(gene_name = ifelse(is.na(gene_name), gene_id, gene_name))

res_mod <- res %>% mutate(lfc = replace(lfc, lfc < -5, -5)) %>% mutate(lfc = replace(lfc, lfc > 5, 5))
show <- as.data.frame(res_mod[1:10, c("lfc", "padj", "gene_name")])

degMA(as.DEGSet(resLFC)) + ggtitle('Before LFC Shrinking')
```

## MA plot

```{r after_lfc_shrink}
degMA(as.DEGSet(resLFCS), limit = 2) + ggtitle('After LFC Shrinking')

```

## Volcano plot

This volcano plot shows the genes that are significantly up- and down-regulated as a result of the analysis comparison. The points highlighted in red are genes that have padj < 0.05 and a log2-fold change > 1. Points in blue have a padj < 0.05 and a log2-fold change < 1 and points in green have a padj > 0.05 and a log2-fold change > 2. Grey points are non-significant. The dashed lines correspond to the cutoff values of log2 foldchance and padj that we have chosen.

```{r volcano_plot, fig.height=6}
# degVolcano(res_mod[,c('lfc', 'padj')], plot_text = show) 
EnhancedVolcano(res_mod,
                lab= res_mod$gene_name, 
                pCutoff = 0.05, 
                selectLab = c(res_sig$gene_name[1:15]),
                FCcutoff = 0.5,
                x = 'lfc',
                y = 'padj', 
                title="Volcano Tumor vs. Normal",
                col=as.vector(colors[c("dark_grey", "light_blue",
                                         "purple", "purple")]),
                subtitle = "", drawConnectors = T,  max.overlaps = Inf) 
```

## Heatmap

```{r heapmap}
### Run pheatmap using the metadata data frame for the annotation
ma=norm_matrix[res_sig$gene_id,]
colma=coldata[,c(column), drop=FALSE]
colors=lapply(colnames(colma), function(c){
  l.col=colors[1:length(unique(colma[[c]]))]
  names(l.col)=unique(colma[[c]])
  l.col
})
names(colors)=colnames(colma)
pheatmap(ma, 
         color = inferno(10), 
         cluster_rows = T, 
         show_rownames = F,
         annotation = colma, 
         annotation_colors = colors,
         border_color = NA, 
         fontsize = 10, 
         scale = "row", 
         fontsize_row = 10, 
         height = 20)
```


## Differentially Expressed Genes

```{r sig_genes_table}
res_sig %>% sanitize_datatable
```

## Plot top 16 genes

```{r top n DEGs, fig.height = 6, fig.width = 8}
n = 16
top_n <- res_sig %>% slice_min(order_by = padj, n = n, with_ties = F) %>% 
  dplyr::select(gene_name, gene_id)
top_n_exp <- norm_matrix %>% as.data.frame() %>% 
  rownames_to_column('gene_id') %>%
  # dplyr::select(-group, -group_name) %>% 
  pivot_longer(!gene_id, names_to = 'sample', values_to = 'log2_expression') %>%
  right_join(top_n, relationship = "many-to-many") %>%
  left_join(coldata, by = 'sample')

ggplot(top_n_exp, aes_string(x = column, y = 'log2_expression')) +
  geom_boxplot(outlier.shape = NA, linewidth=0.5, color="grey") + 
  geom_point() +
  facet_wrap(~gene_name) + 
  ggtitle(str_interp('Expression of Top ${n} DEGs')) +
  theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1))

```

# Pathway Enrichment

From the set of differentially expressed genes and using publicly available information about gene sets involved in biological processes and functions, we can calculate which biological processes and functions are significantly perturbed as a result of the treatment. 

```{r}
universe=res %>% 
  filter(!is.na(padj)) %>% pull(gene_id)
mapping = AnnotationDbi::select(org.Hs.eg.db, universe, 'ENTREZID', 'ENSEMBL')

all_in_life=list(
  msigdbr(species = "human", category = "H") %>% mutate(gs_subcat="Hallmark"),
  msigdbr(species = "human", category = "C2", subcategory = "CP:REACTOME"),
  msigdbr(species = "human", category = "C2", subcategory = "CP:KEGG"),
  msigdbr(species = "human", category = "C2", subcategory = "CP:PID"),
  msigdbr(species = "human", category = "C5", subcategory = "GO:BP"),
  msigdbr(species = "human", category = "C5", subcategory = "GO:MF"),
  msigdbr(species = "human", category = "C5", subcategory = "HPO"),
  msigdbr(species = "human", category = "C3", subcategory = "TFT:GTRD"),
  msigdbr(species = "human", category = "C6") %>% mutate(gs_subcat="Oncogenic")
)

ora_input = res %>% filter(!is.na(padj), padj<0.01, abs(lfc)>0.3) %>% pull(gene_id)
input_entrezid <- AnnotationDbi::select(org.Hs.eg.db, ora_input, 'ENSEMBL', columns = c('ENTREZID', 'SYMBOL'))

total_deg=length(unique(ora_input))/length(unique(mapping$ENTREZID))
pathways_ora_all = lapply(all_in_life, function(p){
  pathway = split(x = p$entrez_gene, f = p$gs_name)
  db_name = paste(p$gs_cat[1], p$gs_subcat[1],sep=":")
  respath <- fora(pathways = pathway, 
                  genes = unique(input_entrezid$ENTREZID),
                  universe = unique(mapping$ENTREZID),
                  minSize  = 15,
                  maxSize  = 500)
  coll_respath = collapsePathwaysORA(respath[order(pval)][padj < 0.1], 
                                     pathway, unique(input_entrezid$ENTREZID), unique(mapping$ENTREZID))
  as_tibble(respath[pathway %in% coll_respath$mainPathways])  %>% 
    mutate(database=db_name, NES=(overlap/size)/(total_deg))
}) %>% bind_rows() %>% 
  mutate(analysis="ORA")
  
ora_tb = pathways_ora_all %>% unnest(overlapGenes) %>%
  group_by(pathway) %>% 
  left_join(mapping, by =c("overlapGenes"="ENTREZID")) %>% 
  dplyr::select(pathway, padj, NES, ENSEMBL, analysis,
                database)

pathways_long = ora_tb

```


```{r pathaways_table}
pathways_ora_all %>% sanitize_datatable()
```


```{r write-files}
counts_norm=norm_matrix %>% as.data.frame() %>% 
  rownames_to_column("gene_id") %>% 
  mutate(comparison = str_interp("${numerator}_vs_${denominator}"))

res_for_writing <- res %>% 
  mutate(comparison = str_interp("${numerator}_vs_${denominator}"))

pathways_for_writing <- pathways_long %>% 
  mutate(comparison = str_interp("${numerator}_vs_${denominator}"))
         
if (!is.null(subset_value)){
  counts_norm <- counts_norm %>% 
    mutate(subset = subset_value)  
  res_for_writing <- res_for_writing %>% 
    mutate(subset = subset_value)
  pathways_for_writing <- pathways_for_writing %>% 
    mutate(subset = subset_value)
}

write_csv(counts_norm, name_expression_fn)
write_csv(res_for_writing, name_deg_fn)
write_csv(pathways_for_writing, name_pathways_fn)
```

# R session

List and version of tools used for the DE report generation.

```{r}
sessionInfo()
```
