---
title: "scRNA MAST"
author: "Harvard Chan Bioinformatics Core"
date: "`r Sys.Date()`"
output:
   html_document:
      code_folding: hide
      df_print: paged
      highlights: pygments
      number_sections: true
      self_contained: true
      theme: default
      toc: true
      toc_float:
         collapsed: true
         smooth_scroll: true
editor_options: 
  chunk_output_type: inline
params:
   project_file: ../information.R
   seurat_obj: "https://github.com/bcbio/bcbioR-test-data/raw/refs/heads/main/singlecell/tiny.rds"
   column: "age"
   contrasts: !r list(c("age", "YOUNG", "OLD"))
   cluster_name: "2"
   resolution_column: "integrated_snn_res.0.4"
   
---

Template developed with materials from https://hbctraining.github.io/main/.


```{r, message=FALSE, warning=FALSE}
# This set up the working directory to this file so all files can be found
# library(rstudioapi)
library(tidyverse)
# setwd(fs::path_dir(getSourceEditorContext()$path))
print(getwd())
stopifnot(R.version$major>= 4) # requires R4
if (compareVersion(R.version$minor,"3.1")<0){
  warning("We recommend >= R4.3.1")}
stopifnot(compareVersion(as.character(BiocManager::version()), "3.16")>=0)
stopifnot(compareVersion(as.character(packageVersion("Seurat")), "5.0.0")>=0)
```


```{r setup, cache=FALSE, message=FALSE, warning=FALSE, echo=FALSE}
knitr::opts_chunk$set(echo = TRUE)
# Load libraries'
library(Seurat)
library(apeglm)
library(SummarizedExperiment)
library(SingleCellExperiment)
library(MAST)
library(tidyverse)
library(data.table)
library(lme4)
library(R.utils)
library(zeallot)
library(ggprism)
library(knitr)

ggplot2::theme_set(theme_prism(base_size = 12))
scale_colour_discrete <- function(...)
  scale_colour_manual(..., values = as.vector(grafify:::graf_palettes[["kelly"]]))

set.seed(1454944673L)
opts_chunk[["set"]](
    audodep = TRUE,
    cache = FALSE,
    cache.lazy = FALSE,
    error = TRUE,
    echo = TRUE,
    fig.height = 5L,
    fig.retina = 2L,
    fig.width = 9.6,
    message = FALSE,
    tidy = TRUE,
    warning = TRUE
)

invisible(list2env(params,environment()))
source(project_file)
```


This code is in this ![](https://img.shields.io/badge/status-alpha-yellow) revision.


# Overview 

-   Project: `r project`
-   PI: `r PI`
-   Analyst: `r analyst`
-   Experiment: `r experiment`
-   Aim: `r aim`

# Dataset load 

```{r load_data, cache = TRUE,message=FALSE, warning=FALSE, echo=FALSE}
# Loading object
if (isUrl(seurat_obj)){
  seurat <- readRDS(url(seurat_obj))  
}else{
  seurat <- readRDS(seurat_obj)
}
DefaultAssay(seurat) <- "RNA"
Idents(object = seurat) <- resolution_column
data_subset <- subset(x = seurat, idents = cluster_name)
sce <- as.SingleCellExperiment(data_subset) 
assay(sce, "log") = log2(counts(sce) + 1)

# Scaling ngenes
cdr = colSums(assay(sce, "log")>0)
colData(sce)$cngeneson = scale(cdr)

# Create new sce object (only 'log' count data)
sce.1 = SingleCellExperiment(assays = list(log = assay(sce, "log")))
colData(sce.1) = colData(sce)

#change to sca
sca = SceToSingleCellAssay(sce.1)
```

```{r meta pre doub}
table(seurat$orig.ident)
```

# MAST analysis 

## Subset genes observed in at least 10% of cells 

```{r}
expressed_genes <- freq(sca) > 0.1
sca_filtered <- sca[expressed_genes, ] 
```

## Prep for mast 

```{r}
cdr2 <- colSums(SummarizedExperiment::assay(sca_filtered)>0)

SummarizedExperiment::colData(sca_filtered)$ngeneson <- scale(cdr2)
SummarizedExperiment::colData(sca_filtered)$orig.ident <-
  factor(SummarizedExperiment::colData(sca_filtered)$orig.ident)
SummarizedExperiment::colData(sca_filtered)[[column]] <-
  factor(SummarizedExperiment::colData(sca_filtered)[[column]])
```

## MAST run with desired contrasts 

```{r,message=FALSE, warning=FALSE, echo=FALSE,cache = TRUE}
comp_name <- levels(SummarizedExperiment::colData(sca_filtered)[[column]])[2]
lrt_name <- paste0(column, comp_name)
formula_touse <- as.formula(paste0("~ ngeneson + (1 | orig.ident) + ", column))

zlmCond <- suppressMessages(MAST::zlm(formula_touse,  sca_filtered, method='glmer',ebayes = F,strictConvergence = FALSE))
summaryCond_column <- suppressMessages(MAST::summary(zlmCond,doLRT=lrt_name))
```


# Result output 

## save conditions used 

```{r}
summary_cond_file = paste0("MAST_RESULTS_",cluster_name,"_", column, ".rds")
saveRDS(summaryCond_column, file = summary_cond_file)
```


## output full mast results 

```{r}
summaryDt_column <- summaryCond_column$datatable
fcHurdle_column <- merge(summaryDt_column[contrast == lrt_name
                                    & component == 'H', .(primerid, `Pr(>Chisq)`)], # This extracts hurdle p-values 
                      summaryDt_column[contrast == lrt_name & component == 'logFC', 
                                    .(primerid, coef, ci.hi, ci.lo)], 
                      by = 'primerid') # This extract LogFC data
fcHurdle_column <- stats::na.omit(as.data.frame(fcHurdle_column))
fcHurdle_column$fdr <- p.adjust(fcHurdle_column$`Pr(>Chisq)`, 'fdr')
to_save_column <- fcHurdle_column

full_res_file = paste0("FULL_MAST_RESULTS_",cluster_name,"_", column, ".csv")
write.table(to_save_column, file=full_res_file, row.names = FALSE, sep=",")
```


## output significant results 

```{r}
fcHurdleSig_column <- merge(fcHurdle_column[fcHurdle_column$fdr < .05,],
                         as.data.table(mcols(sca_filtered)), by = 'primerid')
setorder(fcHurdleSig_column, fdr)
sig_res_file = paste0("SIG_MAST_RESULTS_padj<0.05_",cluster_name,"_", column, ".csv")
write.table(fcHurdleSig_column, file=sig_res_file, row.names = FALSE, sep=",")
```


## Differentially Expressed Genes {.tabset}

```{r sig_genes_table, results='asis'}
DT::datatable(fcHurdleSig_column %>% arrange(fdr))
```


# R session 

List and version of tools used for the QC report generation.

```{r}
sessionInfo()
```

